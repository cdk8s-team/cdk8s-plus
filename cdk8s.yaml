output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
