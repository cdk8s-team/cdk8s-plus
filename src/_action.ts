import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}