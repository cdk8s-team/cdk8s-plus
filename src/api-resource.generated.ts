// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * Represents a resource or collection of resources.
 */
export interface IApiResource {
  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  readonly apiGroup: string;

  /**
   * The name of a resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  readonly resourceType: string;

  /**
   * The unique, namespace-global, name of an object inside the Kubernetes cluster.
   *
   * If this is omitted, the ApiResource should represent all objects of the given type.
   */
  readonly resourceName?: string;
}
/**
 * An API Endpoint can either be a resource descriptor (e.g /pods)
 * or a non resource url (e.g /healthz). It must be one or the other, and not both.
 */
export interface IApiEndpoint {

  /**
   * Return the IApiResource this object represents.
   */
  asApiResource(): IApiResource | undefined;

  /**
   * Return the non resource url this object represents.
   */
  asNonApiResource(): string | undefined;

}

/**
 * Options for `ApiResource`.
 */
export interface ApiResourceOptions {
  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  readonly apiGroup: string;

  /**
   * The name of the resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  readonly resourceType: string;
}

/**
 * Represents information about an API resource type.
 */
export class ApiResource implements IApiResource, IApiEndpoint {
  /**
   * API resource information for Binding.
   */
  public static readonly BINDINGS = new ApiResource({
    apiGroup: '',
    resourceType: 'bindings',
  });

  /**
   * API resource information for ComponentStatus.
   */
  public static readonly COMPONENT_STATUSES = new ApiResource({
    apiGroup: '',
    resourceType: 'componentstatuses',
  });

  /**
   * API resource information for ConfigMap.
   */
  public static readonly CONFIG_MAPS = new ApiResource({
    apiGroup: '',
    resourceType: 'configmaps',
  });

  /**
   * API resource information for Endpoints.
   */
  public static readonly ENDPOINTS = new ApiResource({
    apiGroup: '',
    resourceType: 'endpoints',
  });

  /**
   * API resource information for Event.
   */
  public static readonly EVENTS = new ApiResource({
    apiGroup: '',
    resourceType: 'events',
  });

  /**
   * API resource information for LimitRange.
   */
  public static readonly LIMIT_RANGES = new ApiResource({
    apiGroup: '',
    resourceType: 'limitranges',
  });

  /**
   * API resource information for Namespace.
   */
  public static readonly NAMESPACES = new ApiResource({
    apiGroup: '',
    resourceType: 'namespaces',
  });

  /**
   * API resource information for Node.
   */
  public static readonly NODES = new ApiResource({
    apiGroup: '',
    resourceType: 'nodes',
  });

  /**
   * API resource information for PersistentVolumeClaim.
   */
  public static readonly PERSISTENT_VOLUME_CLAIMS = new ApiResource({
    apiGroup: '',
    resourceType: 'persistentvolumeclaims',
  });

  /**
   * API resource information for PersistentVolume.
   */
  public static readonly PERSISTENT_VOLUMES = new ApiResource({
    apiGroup: '',
    resourceType: 'persistentvolumes',
  });

  /**
   * API resource information for Pod.
   */
  public static readonly PODS = new ApiResource({
    apiGroup: '',
    resourceType: 'pods',
  });

  /**
   * API resource information for PodTemplate.
   */
  public static readonly POD_TEMPLATES = new ApiResource({
    apiGroup: '',
    resourceType: 'podtemplates',
  });

  /**
   * API resource information for ReplicationController.
   */
  public static readonly REPLICATION_CONTROLLERS = new ApiResource({
    apiGroup: '',
    resourceType: 'replicationcontrollers',
  });

  /**
   * API resource information for ResourceQuota.
   */
  public static readonly RESOURCE_QUOTAS = new ApiResource({
    apiGroup: '',
    resourceType: 'resourcequotas',
  });

  /**
   * API resource information for Secret.
   */
  public static readonly SECRETS = new ApiResource({
    apiGroup: '',
    resourceType: 'secrets',
  });

  /**
   * API resource information for ServiceAccount.
   */
  public static readonly SERVICE_ACCOUNTS = new ApiResource({
    apiGroup: '',
    resourceType: 'serviceaccounts',
  });

  /**
   * API resource information for Service.
   */
  public static readonly SERVICES = new ApiResource({
    apiGroup: '',
    resourceType: 'services',
  });

  /**
   * API resource information for MutatingWebhookConfiguration.
   */
  public static readonly MUTATING_WEBHOOK_CONFIGURATIONS = new ApiResource({
    apiGroup: 'admissionregistration.k8s.io',
    resourceType: 'mutatingwebhookconfigurations',
  });

  /**
   * API resource information for ValidatingWebhookConfiguration.
   */
  public static readonly VALIDATING_WEBHOOK_CONFIGURATIONS = new ApiResource({
    apiGroup: 'admissionregistration.k8s.io',
    resourceType: 'validatingwebhookconfigurations',
  });

  /**
   * API resource information for CustomResourceDefinition.
   */
  public static readonly CUSTOM_RESOURCE_DEFINITIONS = new ApiResource({
    apiGroup: 'apiextensions.k8s.io',
    resourceType: 'customresourcedefinitions',
  });

  /**
   * API resource information for APIService.
   */
  public static readonly API_SERVICES = new ApiResource({
    apiGroup: 'apiregistration.k8s.io',
    resourceType: 'apiservices',
  });

  /**
   * API resource information for ControllerRevision.
   */
  public static readonly CONTROLLER_REVISIONS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'controllerrevisions',
  });

  /**
   * API resource information for DaemonSet.
   */
  public static readonly DAEMON_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'daemonsets',
  });

  /**
   * API resource information for Deployment.
   */
  public static readonly DEPLOYMENTS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'deployments',
  });

  /**
   * API resource information for ReplicaSet.
   */
  public static readonly REPLICA_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'replicasets',
  });

  /**
   * API resource information for StatefulSet.
   */
  public static readonly STATEFUL_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'statefulsets',
  });

  /**
   * API resource information for TokenReview.
   */
  public static readonly TOKEN_REVIEWS = new ApiResource({
    apiGroup: 'authentication.k8s.io',
    resourceType: 'tokenreviews',
  });

  /**
   * API resource information for LocalSubjectAccessReview.
   */
  public static readonly LOCAL_SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'localsubjectaccessreviews',
  });

  /**
   * API resource information for SelfSubjectAccessReview.
   */
  public static readonly SELF_SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'selfsubjectaccessreviews',
  });

  /**
   * API resource information for SelfSubjectRulesReview.
   */
  public static readonly SELF_SUBJECT_RULES_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'selfsubjectrulesreviews',
  });

  /**
   * API resource information for SubjectAccessReview.
   */
  public static readonly SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'subjectaccessreviews',
  });

  /**
   * API resource information for HorizontalPodAutoscaler.
   */
  public static readonly HORIZONTAL_POD_AUTOSCALERS = new ApiResource({
    apiGroup: 'autoscaling',
    resourceType: 'horizontalpodautoscalers',
  });

  /**
   * API resource information for CronJob.
   */
  public static readonly CRON_JOBS = new ApiResource({
    apiGroup: 'batch',
    resourceType: 'cronjobs',
  });

  /**
   * API resource information for Job.
   */
  public static readonly JOBS = new ApiResource({
    apiGroup: 'batch',
    resourceType: 'jobs',
  });

  /**
   * API resource information for CertificateSigningRequest.
   */
  public static readonly CERTIFICATE_SIGNING_REQUESTS = new ApiResource({
    apiGroup: 'certificates.k8s.io',
    resourceType: 'certificatesigningrequests',
  });

  /**
   * API resource information for Lease.
   */
  public static readonly LEASES = new ApiResource({
    apiGroup: 'coordination.k8s.io',
    resourceType: 'leases',
  });

  /**
   * API resource information for EndpointSlice.
   */
  public static readonly ENDPOINT_SLICES = new ApiResource({
    apiGroup: 'discovery.k8s.io',
    resourceType: 'endpointslices',
  });

  /**
   * API resource information for FlowSchema.
   */
  public static readonly FLOW_SCHEMAS = new ApiResource({
    apiGroup: 'flowcontrol.apiserver.k8s.io',
    resourceType: 'flowschemas',
  });

  /**
   * API resource information for PriorityLevelConfiguration.
   */
  public static readonly PRIORITY_LEVEL_CONFIGURATIONS = new ApiResource({
    apiGroup: 'flowcontrol.apiserver.k8s.io',
    resourceType: 'prioritylevelconfigurations',
  });

  /**
   * API resource information for IngressClass.
   */
  public static readonly INGRESS_CLASSES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'ingressclasses',
  });

  /**
   * API resource information for Ingress.
   */
  public static readonly INGRESSES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'ingresses',
  });

  /**
   * API resource information for NetworkPolicy.
   */
  public static readonly NETWORK_POLICIES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'networkpolicies',
  });

  /**
   * API resource information for RuntimeClass.
   */
  public static readonly RUNTIME_CLASSES = new ApiResource({
    apiGroup: 'node.k8s.io',
    resourceType: 'runtimeclasses',
  });

  /**
   * API resource information for PodDisruptionBudget.
   */
  public static readonly POD_DISRUPTION_BUDGETS = new ApiResource({
    apiGroup: 'policy',
    resourceType: 'poddisruptionbudgets',
  });

  /**
   * API resource information for PodSecurityPolicy.
   */
  public static readonly POD_SECURITY_POLICIES = new ApiResource({
    apiGroup: 'policy',
    resourceType: 'podsecuritypolicies',
  });

  /**
   * API resource information for ClusterRoleBinding.
   */
  public static readonly CLUSTER_ROLE_BINDINGS = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'clusterrolebindings',
  });

  /**
   * API resource information for ClusterRole.
   */
  public static readonly CLUSTER_ROLES = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'clusterroles',
  });

  /**
   * API resource information for RoleBinding.
   */
  public static readonly ROLE_BINDINGS = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'rolebindings',
  });

  /**
   * API resource information for Role.
   */
  public static readonly ROLES = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'roles',
  });

  /**
   * API resource information for PriorityClass.
   */
  public static readonly PRIORITY_CLASSES = new ApiResource({
    apiGroup: 'scheduling.k8s.io',
    resourceType: 'priorityclasses',
  });

  /**
   * API resource information for CSIDriver.
   */
  public static readonly CSI_DRIVERS = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csidrivers',
  });

  /**
   * API resource information for CSINode.
   */
  public static readonly CSI_NODES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csinodes',
  });

  /**
   * API resource information for CSIStorageCapacity.
   */
  public static readonly CSI_STORAGE_CAPACITIES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csistoragecapacities',
  });

  /**
   * API resource information for StorageClass.
   */
  public static readonly STORAGE_CLASSES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'storageclasses',
  });

  /**
   * API resource information for VolumeAttachment.
   */
  public static readonly VOLUME_ATTACHMENTS = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'volumeattachments',
  });

  /**
   * API resource information for a custom resource type.
   */
  public static custom(options: ApiResourceOptions): ApiResource {
    return new ApiResource(options);
  };

  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  public readonly apiGroup: string;

  /**
   * The name of the resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  public readonly resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }
  private constructor(options: ApiResourceOptions) {
    this.apiGroup = options.apiGroup;
    this.resourceType = options.resourceType;
  }
}

/**
 * Factory for creating non api resources.
 */
export class NonApiResource implements IApiEndpoint {

  public static of(url: string): NonApiResource {
    return new NonApiResource(url);
  }

  private constructor(private readonly nonResourceUrl: string) {};
  public asApiResource(): IApiResource | undefined {
    return undefined;
  }

  public asNonApiResource(): string | undefined {

    return this.nonResourceUrl;
  }
}