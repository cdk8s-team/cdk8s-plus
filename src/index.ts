export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
