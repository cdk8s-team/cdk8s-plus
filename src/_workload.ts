import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}