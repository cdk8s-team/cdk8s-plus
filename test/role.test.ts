import { Testing, ApiObject } from 'cdk8s';
import { Node } from 'constructs';
import * as kplus from '../src';
import { ApiResource } from '../src';

describe('Role', () => {
  test('defaultChild', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const defaultChild = Node.of(
      new kplus.Role(chart, 'pod-reader'),
    ).defaultChild as ApiObject;

    // THEN
    expect(defaultChild.kind).toEqual('Role');

  });

  test('minimal definition', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    new kplus.Role(chart, 'pod-reader');

    // THEN
    expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "Role",
    "metadata": Object {
      "name": "test-pod-reader-c8ec1643",
    },
    "rules": Array [],
  },
]
`);
  });

  test('with a custom resource rule', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.Role(chart, 'pod-reader');
    role.allow(['get', 'watch', 'list'], ApiResource.custom({
      apiGroup: '',
      resourceType: 'pods',
    }));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0]?.rules).toEqual(expect.arrayContaining([
      { apiGroups: [''], resources: ['pods'], verbs: ['get', 'watch', 'list'] },
    ]));

  });

  test('can be allowed read access to a pod and secret', () => {

    // GIVEN
    const chart = Testing.chart();
    const pod = new kplus.Pod(chart, 'my-pod', {
      containers: [
        {
          image: 'nginx',
        },
      ],
    });
    const secret = new kplus.Secret(chart, 'Secret', {
      stringData: {
        key: 'value',
      },
      type: 'kubernetes.io/tls',
    });

    // WHEN
    const role = new kplus.Role(chart, 'pod-reader');
    role.allowRead(pod, secret);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[2].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: [pod.name],
        verbs: ['get', 'list', 'watch'],
      },
      {
        apiGroups: [''],
        resources: ['secrets'],
        resourceNames: [secret.name],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed read access to a mix of resources', () => {

    // GIVEN
    const chart = Testing.chart();
    const pod = new kplus.Pod(chart, 'my-pod', {
      containers: [
        {
          image: 'nginx',
        },
      ],
    });

    // WHEN
    const role = new kplus.Role(chart, 'my-role');
    role.allowRead(kplus.ApiResource.SECRETS, pod);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['secrets'],
        verbs: ['get', 'list', 'watch'],
      },
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: ['test-my-pod-c8a0e457'],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('giving access to a single pod and all pods still gives access to all pods', () => {

    // GIVEN
    const chart = Testing.chart();
    const pod = new kplus.Pod(chart, 'my-pod', {
      containers: [
        {
          image: 'nginx',
        },
      ],
    });

    // WHEN
    const role = new kplus.Role(chart, 'my-role');
    role.allowRead(kplus.ApiResource.PODS, pod);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        verbs: ['get', 'list', 'watch'],
      },
    ]));
    expect(manifest[1].rules).not.toEqual([
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: ['test-my-pod-c8a0e457'],
        verbs: ['get', 'list', 'watch'],
      },
    ]);

  });

  test('can be allowed read access to all pods and secrets in a namespace', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.Role(chart, 'pod-reader');
    role.allowRead(kplus.ApiResource.SECRETS, kplus.ApiResource.PODS);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        verbs: ['get', 'list', 'watch'],
      },
      {
        apiGroups: [''],
        resources: ['secrets'],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed read access to a custom resource type', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.Role(chart, 'pod-reader');
    role.allowRead(kplus.ApiResource.custom({ apiGroup: '', resourceType: 'pods/log' }));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods/log'],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed access to a specific resource and a resource type', () => {

    // GIVEN
    const chart = Testing.chart();

    const pod = new kplus.Pod(chart, 'Pod', { containers: [{ image: 'image' }] });

    // WHEN
    const role = new kplus.Role(chart, 'pod-reader');
    role.allow(['get'], pod, kplus.ApiResource.PODS);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].rules).toEqual(expect.arrayContaining([
      { apiGroups: [''], resourceNames: ['test-pod-c890e1b8'], resources: ['pods'], verbs: ['get'] },
      { apiGroups: [''], resources: ['pods'], verbs: ['get'] },
    ]));

  });

});

describe('ClusterRole', () => {

  test('defaultChild', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const defaultChild = Node.of(
      new kplus.ClusterRole(chart, 'my-cluster-role'),
    ).defaultChild as ApiObject;

    // THEN
    expect(defaultChild.kind).toEqual('ClusterRole');

  });

  test('minimal definition', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    new kplus.ClusterRole(chart, 'my-cluster-role');

    // THEN
    expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRole",
    "metadata": Object {
      "name": "test-my-cluster-role-c86cea4f",
    },
    "rules": Array [],
  },
]
`);

  });

  test('with a custom resource rule', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.ClusterRole(chart, 'pod-reader');
    role.allow(['get', 'watch', 'list'], ApiResource.custom({
      apiGroup: '',
      resourceType: 'pods',
    }));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0]?.rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: [],
        verbs: ['get', 'watch', 'list'],
      },
    ]));

  });

  test('with a custom non-resource rule', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.ClusterRole(chart, 'pod-reader');
    role.allow(['get', 'post'], kplus.NonApiResource.of('/healthz'), kplus.NonApiResource.of('/healthz/*'));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0]?.rules).toEqual(expect.arrayContaining([
      { nonResourceURLs: ['/healthz'], verbs: ['get', 'post'] },
      { nonResourceURLs: ['/healthz/*'], verbs: ['get', 'post'] },
    ]));

  });

  test('can be allowed read access to a pod and secret', () => {

    // GIVEN
    const chart = Testing.chart();
    const pod = new kplus.Pod(chart, 'my-pod', {
      containers: [
        {
          image: 'nginx',
        },
      ],
    });
    const secret = new kplus.Secret(chart, 'Secret', {
      stringData: {
        key: 'value',
      },
      type: 'kubernetes.io/tls',
    });

    // WHEN
    const role = new kplus.ClusterRole(chart, 'my-cluster-role');
    role.allowRead(pod, secret);

    const manifest = Testing.synth(chart);
    expect(manifest[2].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: [pod.name],
        verbs: ['get', 'list', 'watch'],
      },
      {
        apiGroups: [''],
        resources: ['secrets'],
        resourceNames: [secret.name],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed read access to all pods and secrets in the cluster', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.ClusterRole(chart, 'my-cluster-role');
    role.allowRead(kplus.ApiResource.SECRETS, kplus.ApiResource.PODS);

    const manifest = Testing.synth(chart);
    expect(manifest[0].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods'],
        resourceNames: [],
        verbs: ['get', 'list', 'watch'],
      },
      {
        apiGroups: [''],
        resources: ['secrets'],
        resourceNames: [],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed read access to a custom resource type', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.ClusterRole(chart, 'pod-reader');
    role.allowRead(kplus.ApiResource.custom({ apiGroup: '', resourceType: 'pods/log' }));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].rules).toEqual(expect.arrayContaining([
      {
        apiGroups: [''],
        resources: ['pods/log'],
        resourceNames: [],
        verbs: ['get', 'list', 'watch'],
      },
    ]));

  });

  test('can be allowed access to a specific resource, a resource type, and non resource endpoints', () => {

    // GIVEN
    const chart = Testing.chart();

    const pod = new kplus.Pod(chart, 'Pod', { containers: [{ image: 'image' }] });

    // WHEN
    const role = new kplus.ClusterRole(chart, 'pod-reader');
    role.allow(['get'], pod, kplus.ApiResource.PODS, kplus.NonApiResource.of('/healthz'));

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[1].rules).toEqual(expect.arrayContaining([
      { apiGroups: [''], resourceNames: ['test-pod-c890e1b8'], resources: ['pods'], verbs: ['get'] },
      { apiGroups: [''], resourceNames: [], resources: ['pods'], verbs: ['get'] },
      { nonResourceURLs: ['/healthz'], verbs: ['get'] },
    ]));

  });

  test('can be aggregated', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role1 = new kplus.ClusterRole(chart, 'pod-reader');
    role1.allow(['get', 'watch', 'list'], ApiResource.PODS);

    const role2 = new kplus.ClusterRole(chart, 'secrets-reader');
    role2.allow(['get', 'watch', 'list'], ApiResource.SECRETS);

    const combined = new kplus.ClusterRole(chart, 'combined-role');
    combined.combine(role1);
    combined.combine(role2);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].metadata.labels).toMatchInlineSnapshot(`
Object {
  "cdk8s.cluster-role/aggregate-to-test-combined-role-c8db37c0": "true",
}
`);
    expect(manifest[1].metadata.labels).toMatchInlineSnapshot(`
Object {
  "cdk8s.cluster-role/aggregate-to-test-combined-role-c8db37c0": "true",
}
`);
    expect(manifest[2].aggregationRule).toMatchInlineSnapshot(`
Object {
  "clusterRoleSelectors": Array [
    Object {
      "matchLabels": Object {
        "cdk8s.cluster-role/aggregate-to-test-combined-role-c8db37c0": "true",
      },
    },
  ],
}
`);

  });

  test('custom aggregation labels can be added', () => {

    // GIVEN
    const chart = Testing.chart();

    // WHEN
    const role = new kplus.ClusterRole(chart, 'secrets-reader', {
      aggregationLabels: {
        // adds these permissions to the "view" default role
        'rbac.authorization.k8s.io/aggregate-to-view': 'true',
      },
    });
    // add these permissions to the "admin" default role
    role.aggregate('rbac.authorization.k8s.io/aggregate-to-admin', 'true');
    role.allow(['get', 'watch', 'list'], ApiResource.SECRETS);

    // THEN
    const manifest = Testing.synth(chart);
    expect(manifest[0].aggregationRule).toMatchInlineSnapshot(`
Object {
  "clusterRoleSelectors": Array [
    Object {
      "matchLabels": Object {
        "rbac.authorization.k8s.io/aggregate-to-admin": "true",
        "rbac.authorization.k8s.io/aggregate-to-view": "true",
      },
    },
  ],
}
`);

  });

});