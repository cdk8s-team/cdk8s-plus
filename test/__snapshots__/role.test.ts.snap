// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterRole can grant permissions on imported 1`] = `
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "Role",
    "metadata": Object {
      "name": "test-role-c8eba390",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "rbac.authorization.k8s.io",
        ],
        "resourceNames": Array [
          "r2",
        ],
        "resources": Array [
          "clusterroles",
        ],
        "verbs": Array [
          "get",
          "list",
          "watch",
        ],
      },
    ],
  },
]
`;

exports[`Role can grant permissions on imported 1`] = `
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "Role",
    "metadata": Object {
      "name": "test-role-c8eba390",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "rbac.authorization.k8s.io",
        ],
        "resourceNames": Array [
          "r2",
        ],
        "resources": Array [
          "roles",
        ],
        "verbs": Array [
          "get",
          "list",
          "watch",
        ],
      },
    ],
  },
]
`;
